File: src/bloatit/htmlrenderer/htmltools.py
# -*- coding: utf-8 -*-

# Copyright (C) 2010 BloatIt.
#
# This file is part of BloatIt.
#
# BloatIt is free software: you can redistribute it and/or modify
# it under the terms of the GNU Affero General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# BloatIt is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
# GNU Affero General Public License for more details.
#
# You should have received a copy of the GNU Affero General Public License
# along with BloatIt. If not, see <http://www.gnu.org/licenses/>.

class HtmlTools:
    
   
    @classmethod
    def generate_logo(cls):
        return '<span class="logo_bloatit"><span class="logo_bloatit_bloat">Bloat</span><span class="logo_bloatit_it">It</span></span>'

    @classmethod
    def generate_link(cls, session, text, link_page):
        # @type link_page PageContent
        return '<a href="/'+session.get_language().get_code()+'/'+link_page.get_code()+'">'+text+'</a>'

    @classmethod
    def generate_action_link(cls, session, text, link_action):
        # @type link_action Action
        return '<a href="/'+session.get_language().get_code()+'/action/'+link_action.get_code()+'">'+text+'</a>'

    @classmethod
    def generate_url(cls, session, url_page):
        # @type url_page PageContent
        return '/'+session.get_language().get_code()+'/'+url_page.get_code()


    def cut_number(number):
        result = number
        if len(result) > 2:
                if result[1] == ".":
                    if result[2] == "0":
                        result = result[0]
                    else:
                        result = result[:3]
                elif result[2] == ".":
                    result = result[:2]
                else:
                    result = result[:3]
        return result

    @classmethod
    def compress_karma(cls, karma):
        abs_karma = abs(karma)

        if abs_karma < 1000:
            result = str(abs_karma)
        elif abs_karma < 1000000:
            result = str(abs_karma/1000)
            result = cls.cut_number(result)+"k"
        elif abs_karma < 1000000000:
            result = str(abs_karma/1000000)
            result = cls.cut_number(result)+"M"
        elif abs_karma < 1000000000000:
            result = str(abs_karma/1000000000)
            result = cls.cut_number(result)+"T"
        else:
            result = "∞"

        
        if(karma >= 0):
            return result
        else:
            return "-"+result


    
        

File: src/main/java/com/bloatit/web/htmlrenderer/HtmlTools.java
/*
 * Copyright (C) 2010 BloatIt.
 * 
 * This file is part of BloatIt.
 * 
 * BloatIt is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 * 
 * BloatIt is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU Affero General Public License for more details.
 * 
 * You should have received a copy of the GNU Affero General Public License
 * along with BloatIt. If not, see <http://www.gnu.org/licenses/>.
 */
package com.bloatit.web.htmlrenderer;

import com.bloatit.web.server.Action;
import com.bloatit.web.server.Page;
import com.bloatit.web.server.Session;

public class HtmlTools {

    public static String generateLink(Session session, String displayedText, Page linkPage) {
        return "<a href=\"/" + session.getLanguage().getCode() + "/" + linkPage.getCode() + "\">" + displayedText + "</a>";
    }

    public static String generateActionLink(Session session, String text, Action linkAction) {
        return "<a href=\"/" + session.getLanguage().getCode() + "/action/" + linkAction.getCode() + "\">" + text + "</a>";
    }

    public static String generateUrl(Session session, Page urlPage) {
        return "/" + session.getLanguage().getCode() + "/" + urlPage.getCode();
    }

    public static String generateLogo() {
        return "<span class='logo_bloatit'><span class='logo_bloatit_bloat'>Bloat</span><span class='logo_bloatit_it'>It</span></span>";
    }

    public static String escapeUrlString(String str) {
        // TODO return (urllib.parse.quote_plus(str)).replace('+','_');
        return str;
    }

    public static String unescapeUrlString(String str) {
        // TODO return (urllib.parse.quote_plus(str)).replace('+','_');
        return str;
    }

    /**
     * Compress karma and make it easier to display
     * 
     * Example of results :
     * 1 = 1
     * 100 = 100
     * 1 000 = 1k
     * 100 000 = 100k
     * 1 000 000 = 1M
     * 100 000 000 = 100M
     * 1 000 000 000 = 1T
     * 100 000 000 000 = 100T
     * 1 000 000 000 000 = ∞
     * 
     * @param karma the karma value to compress
     * @return the compressed String to display
     */
    public static String compressKarma(long karma) {
        Double abs_karma = new Double(Math.abs(karma));
        String result = "";
        if (abs_karma < 1000) {
            result = cutNumber(abs_karma.toString());
        } else if (abs_karma < 1000000d) {
            result = cutNumber(new Double(abs_karma / 1000d).toString()) + "k";// TODO why not 'K' ?
        } else if (abs_karma < 1000000000d) {
            result = cutNumber(new Double(abs_karma / 1000000d).toString()) + "M";

        } else if (abs_karma < 1000000000000d) {
            result = cutNumber(new Double(abs_karma / 1000000000d).toString()) + "T";
        } else {

            result = "∞";
        }
        if (karma >= 0) {
            return result;

        } else {
            return "-" + result;

        }
    }

    public static String cutNumber(String karma) {
        String result = karma;
        if (result.length() > 2) {
            if (result.charAt(1) == '.') {
                if (result.charAt(2) == '0') {
                    result = result.substring(0, 1);
                } else {
                    result = result.substring(0, 3);
                }
            } else if (result.charAt(2) == '.') {
                result = result.substring(0, 2);
            } else {

                result = result.substring(0, 3);
            }
        }
        return result;
    }
}
